portia_server/portia_api/resources/projects.py

```diff
@@ -98,39 +109,107 @@ def status(self, *args, **kwargs):
 
     @detail_route(methods=['put', 'patch', 'post'])
     def publish(self, *args, **kwargs):
-        # manager = self.project_spec
-        # if not hasattr(manager.pm, 'publish_project'):
-        #     raise JsonApiFeatureNotAvailableError()
-        # project_id = manager.project_name
-        # if not self.get_project_changes():
-        #     raise BadRequest('The project is up to date')
-        # publish_status = json.loads(
-        #     manager.pm.publish_project(project_id,
-        #                                self.data.get('force', False)))
-        # if publish_status['status'] == 'conflict':
-        #     raise BaseError(409, 'A conflict has occurred in this project',
-        #                     'You must resolve the conflict for the project to be'
-        #                     ' successfully published')
+        if not self.storage.version_control and hasattr(self.storage, 'repo'):
+            raise JsonApiFeatureNotAvailableError()
+
+        if not self.get_project_changes():
+            raise JsonApiBadRequestError('You have no changes to publish')
 
+        force = self.query.get('force', False)
+        branch = self.storage.branch
+        published = self.storage.repo.publish_branch(branch, force=force)
+        if not published:
+            return JsonApiConflictError(
+                'A conflict occurred when publishing your changes.'
+                'You must resolve the conflict before the project can be '
+                'published.')
+        self.deploy()
+        self.storage.repo.delete_branch(branch)
         response = self.retrieve()
-        data = OrderedDict()
-        # data.update({
-        #     'meta': manager.pm._schedule_data(project_id=project_id)
-        # })
-        data.update(response.data)
-        return Response(data, status=HTTP_200_OK)
+        return Response(response.data, status=HTTP_200_OK)
 
     @detail_route(methods=['put', 'patch', 'post'])
     def reset(self, *args, **kwargs):
-        # manager = self.project_spec
-        # if not hasattr(manager.pm, 'discard_changes'):
-        #     raise JsonApiFeatureNotAvailableError()
-        # project_id = manager.project_name
-        # if not self.get_project_changes():
-        #     raise BadRequest('There are no changes to discard')
-        # manager.pm.discard_changes(project_id)
+        if not self.storage.version_control and hasattr(self.storage, 'repo'):
+            raise JsonApiFeatureNotAvailableError()
+        branch = self.storage.branch
+        master = self.storage.repo.refs['refs/heads/master']
+        self.storage.repo.refs['refs/heads/%s' % branch] = master
         return self.retrieve()
 
+    @detail_route(methods=['post'])
+    def copy(self, *args, **kwargs):
```

Comment:
Can we break up this function? Something like `validation`, `error handling`, `copying logic`, and the returned data?